File: C++/029. Divide Two Integers.cpp
class Solution {
public:
    int divide(int dividend, int divisor) {
        if(dividend>INT_MAX || dividend<INT_MIN || divisor>INT_MAX || divisor<INT_MIN){return INT_MAX;}
        long long res=0;
        unsigned long val1,val2,temp1,temp2;
        val1 = (dividend>=0)? dividend:-(long long)dividend; 
        val2 = (divisor>=0)? divisor:-(long long)divisor; 
        temp1 = val1;
        temp2 = val2;
        long long c = 1;
        if(val1 < val2){return 0;}
        while(val1 >= temp2)        
        {
            temp2 = temp2 << 1;
            c= c << 1;
        }
        while(temp1>=val2)
        {
            while(temp1>=temp2)
            {
                temp1 -= temp2;
                res += c; 
            }
            temp2 = temp2 >> 1;
            c = c>>1;
        }
        
        if((dividend<0 && divisor>0) || (dividend>0 && divisor<0))
        {
            res = 0-res;
        }
        if(res>INT_MAX ||res<INT_MIN)
        {
            return INT_MAX;
        }
        else
        {
            return (int)res;
        }
    }
};

File: python/029. Divide Two Integers.py
class Solution(object):
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        ans = 0
        if (dividend > 0 and divisor < 0) or (dividend < 0 and divisor > 0):
            sign = -1
        else:
            sign = 1
        dividend, divisor = abs(dividend), abs(divisor)
        val1, val2 = dividend, divisor
        base = 1
        while val2 < val1:
            val2 <<= 1
            base <<= 1
        while val2 >= divisor:
            if val1 >= val2:
                ans += base
                val1 -= val2
            val2 >>= 1
            base >>= 1
        ans *= sign
        lowerBound = - 2 ** 31
        upperBound = 2 ** 31 -1
        if ans < lowerBound or ans > upperBound:
            return upperBound
        else:
            return ans